### ISSUE TITLE
AbortController is missing

### ISSUE THREAD

**Issue by giacomocerquone** [facebook/react-native#18115]
### Is this a bug report?

Yes. AbortController is a new spec (you can find info here: https://developers.google.com/web/updates/2017/09/abortable-fetch)

### Have you read the [Contributing Guidelines](https://facebook.github.io/react-native/docs/contributing.html)?

Yes

### Environment

```
Environment:
  OS: Linux 4.13
  Node: 8.9.4
  Yarn: 1.3.2
  npm: 5.6.0
  Watchman: Not Found
  Xcode: N/A
  Android Studio: Not Found

Packages: (wanted => installed)
  react: ^16.0.0 => 16.2.0
  react-native: ^0.53.0 => 0.53.0
```

### Steps to Reproduce

1. Istantiate an AbortController
2. Write a fetch request passing the signal
3. Try to abort the fetch request

### Expected Behavior

The requested behaviour is to abort the fetch request.

### Actual Behavior

Nothing happens

### Reproducible Demo

Something like this:

```
const controller = new AbortController();
const signal = controller.signal;

setTimeout(() => controller.abort(), 5000);

fetch(url, { signal }).then(response => {
  return response.text();
}).then(text => {
  console.log(text);
});
```

My question is a technical one: is the react native core that needs to be updated in order to support this? Or is it something that babel can patch with a new version?

**Comment by dooboolab** [comment]
Having same issue. This should be in react-native too.

**Comment by chirag04** [comment]
might be worth waiting for https://github.com/github/fetch/pull/572 to merge.

**Comment by giacomocerquone** [comment]
Very good, at least I should be able to use[ this polyfill](https://github.com/mo/abortcontroller-polyfill/) right now in react native right?

**Comment by chirag04** [comment]
I haven’t used it but should be able to use it. Might be worth posting here what you find out

**Comment by giacomocerquone** [comment]
Great, it works!
I'm not on a create react app, but I've just followed the instruction for it (I'm actually on a react-native project **not even on a create react native app**

So installed the module, imported it like this:

```javascript
import 'abortcontroller-polyfill';
```

and then used it like this:

```javascript
const AbortController = window.AbortController;
const controller = new AbortController()
const signal = controller.signal
```

I'd not close the issue, because we're still waiting for the official spec.
Thank you @chirag04 for linking me the right resource :relaxed:

**Comment by sibelius** [comment]
@giacomocerquone the polyfill is working on react-native (https://github.com/facebook/react-native/issues/18115#issuecomment-368978604)

do u have an example of it?

**Comment by giacomocerquone** [comment]
Just doing what I wrote in the previous comment should be fine. All the code then would be equals to the one used in the browsers like the piece showed on mozilla doc: https://developer.mozilla.org/en-US/docs/Web/API/AbortController/abort

I've actually a production app where I'm using this. I can't share directly the code of course, but I can re-arrange an example if it is strictly needed :)

**Comment by sibelius** [comment]
it works, but won't really cancel the request, it will just throw an exception 

tks

**Comment by giacomocerquone** [comment]
Basically it just rejects the promise.
It's true that it won't cancel the request (we should also specify precisely what we mean when we write 'cancel'), but as far as I'm concerned, my interest was to not let the request hanging and to 'close the door' to all the responses that would have come after the abort call.

**Comment by coolboiime** [comment]
I just tried it today with the polyfills, and it doesn’t work.

**Comment by TimurAsayonok** [comment]
Same problem.

Try to cancel fetch after timeout (5 sec), for simulating bad connection I'm using Network link conditioner

**Comment by m-vdb** [comment]
```bash
$ npm ls whatwg-fetch
└─┬ react-native@0.57.0
  └─┬ fbjs@0.8.17
    └─┬ isomorphic-fetch@2.2.1
      └── whatwg-fetch@3.0.0
```

**Comment by arthurfranca** [comment]
I think the polyfill works as expected on RN 0.57. But sometimes, the abort event is received after the fetch promise has already resolved [and won the Promise.race](https://github.com/mo/abortcontroller-polyfill/blob/master/src/abortableFetch.js#L77).
Probably because of [using setTimeout](https://github.com/mo/abortcontroller-polyfill/blob/master/src/abortcontroller.js#L28) that places the dispatch event on a queue.

One way to get around this is checking for `controller.signal.aborted` flag if you are no longer interested in the response. (e.g. the component is unmounted)

**Comment by hyochan** [comment]
How do you solve the `jest` mocking when using @giacomocerquone solution?
![image](https://user-images.githubusercontent.com/27461460/50156280-c0791980-0311-11e9-9116-be027bae4f1d.png)

I've tried to mock `AbortController` like below and won't work.
```
export default class AbortController {
  constructor() {
    this.controller = new AbortController();
  }
}
```

**Comment by Jyrno42** [comment]
> Apparently, this issue should not happen with react-native 0.57 since `whatwg-fetch` was remove with [this commit](https://github.com/facebook/react-native/commit/122b3791ede095345f44666691aa9ce5aa7f725a#diff-b9cfc7f2cdf78a7f4b91a753d10865a2) but I'm not 100% sure.

The polyfill in that commit does not contain any logic for Abort controllers so I do not think it works on latest rn as well (since the vendored file is based on 1.0.0 of whatwg-fetch).

**Comment by TimurAsayonok** [comment]
Any updates of this?
Have same problem. Try to use new version of `whatwg-fetch` 3.0.0 with Aborting fetch, and it's work but, when we create new build it will crash. Because they change code in 3.0.0 for browsers.

@chmartinez can you give example how we can you `abortableFetch` from `abortcontroller-polyfill`

**Comment by TimurAsayonok** [comment]
So, I try to use [cross-fetch](https://github.com/lquixada/cross-fetch) instead latest version of whatwg-fetch

**Comment by giacomocerquone** [comment]
I wrote an article for all the people that were asking me to shed some light about this matter: https://blog.giacomocerquone.com/aborting-fetch-react-native/

In the next one I will give you some ready to use code to reject requests, but this time instead of giving you code I wanted to explain what is going on and how and why is useless to use polyfills in react native as of now.

Enjoy, hope I did a good thing!
As I wrote in the post, in the next article I will share with you my simple and short piece of code that I use to make requests that implements fetch rejection with few lines of code among other things

**Comment by atkit** [comment]
Seems like work on AbortController for RN was already done.
https://github.com/facebook/react-native/commit/5e36b0c6eb2494cefd11907673aa018831526750

**Comment by raunaqss** [comment]
@atkit Yes, work on this has been done, however, a comment on the issue related to this merge says that this work is focussed towards RN 0.60+, hence not support on previous versions.

**Comment by giacomocerquone** [comment]
@raunaqss as I've already said, an implementation from facebook was needed in order to make an abortion work, hence old version cannot be supported. Why the downvote?

**Comment by JordanRichards-hx** [comment]
Support for this was included in the new version; 0.60, which was released yesterday.

https://github.com/react-native-community/releases/blob/master/CHANGELOG.md#060

**Comment by SaeedZhiany** [comment]
@JordanRichards-hx 
I just created a fresh Rn project (0.60.5) with typescript. I see nothing added to fetch's `RequestInfo` and `RequestInit` interfaces. is the typescript definition files outdated or we must use RN's fetch API differently? I just confused about how I use the feature.

**Comment by atkit** [comment]
I found an example here:
https://github.com/facebook/react-native/commit/5e36b0c#diff-3e169de4973928b8c4e90923cac0651aR22

And I checked that it is possible to create instance of AbortController but yes @types/react-native still have no information about AbortController so typescript shows an error. I think it is good opportunity to contribute to @types/react-native

**Comment by SaeedZhiany** [comment]
Is `AbortController` an RN built-in class? or I should install it from [abort-controller repository](https://github.com/mysticatea/abort-controller)

I'm asking because of the typescript issue. maybe it's possible to export a utility function to create it for the user and so there is no need to install the lib in the main project

**Comment by atkit** [comment]
As I can see from codebase it is installed as dependancy to react-native, but typing information is not yet in place for some reason. Typing information lives outside of react-native in repo for all typescript types. As it is a global class for React Native this typing needs to be provided. 

I did not start using this AbortController in my apps yet, but I'm planning to implement timeout and cancel operations at some point. I checked quickly it does work and it cancels operation with special type of Error.
